File: java/src/main/java/ai/buble/sdk/http/BubleHttpClient.java
package ai.buble.sdk.http;

import ai.buble.sdk.BubleApiException;
import ai.buble.sdk.BubleException;
import ai.buble.sdk.BubleTimeoutException;
import ai.buble.sdk.RequestOptions;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Internal HTTP adapter used by SDK resources.
 */
public final class BubleHttpClient {
    private final String apiKey;
    private final String baseUrl;
    private final Duration timeout;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final Map<String, String> headers;

    public BubleHttpClient(
            String apiKey,
            String baseUrl,
            Duration timeout,
            HttpClient httpClient,
            ObjectMapper objectMapper,
            Map<String, String> headers
    ) {
        this.apiKey = apiKey;
        this.baseUrl = trimTrailingSlash(baseUrl);
        this.timeout = timeout;
        this.httpClient = httpClient;
        this.objectMapper = objectMapper;
        this.headers = Collections.unmodifiableMap(new LinkedHashMap<String, String>(headers));
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public ObjectMapper getObjectMapper() {
        return objectMapper;
    }

    public <T> T get(String path, TypeReference<T> type) {
        return request("GET", path, null, type, RequestOptions.none());
    }

    public <T> T get(String path, TypeReference<T> type, RequestOptions options) {
        return request("GET", path, null, type, options);
    }

    public <T> T post(String path, Object body, TypeReference<T> type) {
        return request("POST", path, body, type, RequestOptions.none());
    }

    public <T> T post(String path, Object body, TypeReference<T> type, RequestOptions options) {
        return request("POST", path, body, type, options);
    }

    public HttpResponse<InputStream> stream(String path, Object body, RequestOptions options) {
        byte[] json = encodeJson(body == null ? Collections.emptyMap() : body);
        HttpRequest request = baseRequest(path, options)
                .header("Content-Type", "application/json")
                .setHeader("Accept", "text/event-stream")
                .POST(HttpRequest.BodyPublishers.ofByteArray(json))
                .build();
        return sendStream(request);
    }

    public <T> T multipart(String path, MultipartBody body, TypeReference<T> type, RequestOptions options) {
        HttpRequest request = baseRequest(path, options)
                .header("Content-Type", "multipart/form-data; boundary=" + body.getBoundary())
                .POST(body.toBodyPublisher())
                .build();
        HttpResponse<String> response = send(request);
        return parseSuccessful(response, type);
    }

    private <T> T request(String method, String path, Object body, TypeReference<T> type, RequestOptions options) {
        HttpRequest.BodyPublisher publisher;
        HttpRequest.Builder builder = baseRequest(path, options);

        if (body == null) {
            publisher = HttpRequest.BodyPublishers.noBody();
        } else {
            publisher = HttpRequest.BodyPublishers.ofByteArray(encodeJson(body));
            builder.header("Content-Type", "application/json");
        }

        HttpRequest request = builder.method(method, publisher).build();
        HttpResponse<String> response = send(request);
        return parseSuccessful(response, type);
    }

    private HttpRequest.Builder baseRequest(String path, RequestOptions options) {
        if (apiKey == null || apiKey.isEmpty()) {
            throw new BubleException("Missing Buble API key. Pass apiKey or set BUBLE_API_KEY.");
        }
        RequestOptions resolved = options == null ? RequestOptions.none() : options;
        HttpRequest.Builder builder = HttpRequest.newBuilder(resolve(path, resolved.getQuery()))
                .timeout(timeout)
                .header("Authorization", "Bearer " + apiKey)
                .header("Accept", "application/json");
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            builder.header(entry.getKey(), entry.getValue());
        }
        for (Map.Entry<String, String> entry : resolved.getHeaders().entrySet()) {
            builder.setHeader(entry.getKey(), entry.getValue());
        }
        return builder;
    }

    private HttpResponse<String> send(HttpRequest request) {
        try {
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw apiException(response.statusCode(), response.body());
            }
            return response;
        } catch (java.net.http.HttpTimeoutException e) {
            throw new BubleTimeoutException("Buble API request timed out after " + timeout + ".", timeout, e);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new BubleException("Buble API request was interrupted.", e);
        }
    }

    private HttpResponse<InputStream> sendStream(HttpRequest request) {
        try {
            HttpResponse<InputStream> response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                String responseBody = new String(response.body().readAllBytes(), StandardCharsets.UTF_8);
                throw apiException(response.statusCode(), responseBody);
            }
            return response;
        } catch (java.net.http.HttpTimeoutException e) {
            throw new BubleTimeoutException("Buble API request timed out after " + timeout + ".", timeout, e);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new BubleException("Buble API request was interrupted.", e);
        }
    }

    private <T> T parseSuccessful(HttpResponse<String> response, TypeReference<T> type) {
        if (response.statusCode() == 204 || type == null || response.body() == null || response.body().isEmpty()) {
            return null;
        }
        try {
            return objectMapper.readValue(response.body(), type);
        } catch (JsonProcessingException e) {
            throw new BubleException("Failed to parse Buble API response.", e);
        }
    }

    private byte[] encodeJson(Object body) {
        try {
            return objectMapper.writeValueAsBytes(body);
        } catch (JsonProcessingException e) {
            throw new BubleException("Failed to encode Buble API request body.", e);
        }
    }

    private BubleApiException apiException(int statusCode, String responseBody) {
        String message = responseBody;
        String code = null;
        JsonNode details = null;
        if (message == null || message.isEmpty()) {
            message = "Buble API request failed with status " + statusCode + ".";
        }
        try {
            JsonNode root = objectMapper.readTree(responseBody);
            JsonNode error = root.path("error");
            if (error.isObject()) {
                if (error.hasNonNull("message")) {
                    message = error.get("message").asText();
                }
                if (error.hasNonNull("code")) {
                    code = error.get("code").asText();
                }
                if (error.has("details")) {
                    details = error.get("details");
                }
            }
        } catch (Exception ignored) {
            // Use the raw response body when the server did not return JSON.
        }
        return new BubleApiException(statusCode, code, message, details, responseBody);
    }

    private URI resolve(String path, Map<String, String> query) {
        String normalizedPath = path.startsWith("/") ? path : "/" + path;
        StringBuilder url = new StringBuilder(baseUrl).append(normalizedPath);
        boolean hasQuery = normalizedPath.contains("?");
        for (Map.Entry<String, String> entry : query.entrySet()) {
            url.append(hasQuery ? '&' : '?');
            hasQuery = true;
            url.append(urlEncode(entry.getKey())).append('=').append(urlEncode(entry.getValue()));
        }
        return URI.create(url.toString());
    }

    public static String encodePathSegment(String value) {
        return urlEncode(value);
    }

    public static String encodeModelPath(String model) {
        String[] parts = model.split("/", -1);
        StringBuilder encoded = new StringBuilder();
        for (int i = 0; i < parts.length; i++) {
            if (i > 0) {
                encoded.append('/');
            }
            encoded.append(urlEncode(parts[i]));
        }
        return encoded.toString();
    }

    private static String urlEncode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String trimTrailingSlash(String value) {
        String out = value == null || value.isEmpty() ? "https://buble.ai" : value;
        while (out.endsWith("/")) {
            out = out.substring(0, out.length() - 1);
        }
        return out;
    }
}

File: dotnet/src/Buble.Sdk/Http/BubleHttpClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Buble.Sdk.Http;

internal sealed class BubleHttpClient : IDisposable
{
    private readonly string _apiKey;
    private readonly TimeSpan _timeout;
    private readonly HttpClient _httpClient;
    private readonly IReadOnlyDictionary<string, string> _headers;

    internal BubleHttpClient(
        string apiKey,
        string baseUrl,
        TimeSpan timeout,
        HttpClient httpClient,
        IDictionary<string, string> headers)
    {
        _apiKey = apiKey;
        BaseUrl = TrimTrailingSlash(baseUrl);
        _timeout = timeout;
        _httpClient = httpClient;
        _headers = new Dictionary<string, string>(headers, StringComparer.OrdinalIgnoreCase);
    }

    internal string BaseUrl { get; }

    internal Task<T?> GetAsync<T>(string path, RequestOptions? options = null, CancellationToken cancellationToken = default)
    {
        return SendJsonAsync<T>(HttpMethod.Get, path, null, options, cancellationToken);
    }

    internal Task<T?> PostAsync<T>(string path, object? body, RequestOptions? options = null, CancellationToken cancellationToken = default)
    {
        return SendJsonAsync<T>(HttpMethod.Post, path, body, options, cancellationToken);
    }

    internal async Task<T?> MultipartAsync<T>(
        string path,
        MultipartFormDataContent body,
        RequestOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        using var request = CreateRequest(HttpMethod.Post, path, options);
        request.Content = body;
        using var response = await SendAsync(request, HttpCompletionOption.ResponseContentRead, cancellationToken).ConfigureAwait(false);
        return await ParseSuccessfulAsync<T>(response).ConfigureAwait(false);
    }

    internal async IAsyncEnumerable<string> StreamLinesAsync(
        string path,
        object? body,
        RequestOptions? options = null,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        using var request = CreateRequest(HttpMethod.Post, path, options);
        var json = JsonSerializer.Serialize(body ?? new Dictionary<string, object?>(), BubleJson.Options);
        request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        request.Headers.Accept.Clear();
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));

        using var response = await SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
        var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
        using var reader = new StreamReader(stream, Encoding.UTF8);

        while (!reader.EndOfStream)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var line = await reader.ReadLineAsync().ConfigureAwait(false);
            if (line is not null)
            {
                yield return line;
            }
        }
    }

    internal static string EncodePathSegment(string value) => Uri.EscapeDataString(value);

    internal static string EncodeModelPath(string model)
    {
        return string.Join("/", model.Split('/').Select(Uri.EscapeDataString));
    }

    private async Task<T?> SendJsonAsync<T>(
        HttpMethod method,
        string path,
        object? body,
        RequestOptions? options,
        CancellationToken cancellationToken)
    {
        using var request = CreateRequest(method, path, options);
        if (body is not null)
        {
            var json = JsonSerializer.Serialize(body, BubleJson.Options);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        using var response = await SendAsync(request, HttpCompletionOption.ResponseContentRead, cancellationToken).ConfigureAwait(false);
        return await ParseSuccessfulAsync<T>(response).ConfigureAwait(false);
    }

    private HttpRequestMessage CreateRequest(HttpMethod method, string path, RequestOptions? options)
    {
        var resolved = options ?? RequestOptions.None;
        var request = new HttpRequestMessage(method, Resolve(path, resolved.Query));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        foreach (var header in _headers)
        {
            request.Headers.TryAddWithoutValidation(header.Key, header.Value);
        }

        foreach (var header in resolved.Headers)
        {
            request.Headers.Remove(header.Key);
            request.Headers.TryAddWithoutValidation(header.Key, header.Value);
        }

        return request;
    }

    private async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request,
        HttpCompletionOption completionOption,
        CancellationToken cancellationToken)
    {
        using var timeoutCts = new CancellationTokenSource(_timeout);
        using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);

        try
        {
            var response = await _httpClient.SendAsync(request, completionOption, linkedCts.Token).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                response.Dispose();
                throw CreateApiException((int)response.StatusCode, body);
            }

            return response;
        }
        catch (TaskCanceledException ex) when (!cancellationToken.IsCancellationRequested)
        {
            throw new BubleTimeoutException($"Buble API request timed out after {_timeout}.", _timeout, ex);
        }
    }

    private static async Task<T?> ParseSuccessfulAsync<T>(HttpResponseMessage response)
    {
        if (response.StatusCode == HttpStatusCode.NoContent)
        {
            return default;
        }

        var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(body))
        {
            return default;
        }

        try
        {
            return JsonSerializer.Deserialize<T>(body, BubleJson.Options);
        }
        catch (JsonException ex)
        {
            throw new BubleException("Failed to parse Buble API response.", ex);
        }
    }

    private static BubleApiException CreateApiException(int statusCode, string? responseBody)
    {
        var message = string.IsNullOrWhiteSpace(responseBody)
            ? $"Buble API request failed with status {statusCode}."
            : responseBody!;
        string? code = null;
        JsonNode? details = null;

        try
        {
            var root = JsonNode.Parse(responseBody ?? string.Empty);
            var error = root?["error"];
            if (error is not null)
            {
                message = error["message"]?.GetValue<string>() ?? message;
                code = error["code"]?.GetValue<string>();
                details = error["details"];
            }
        }
        catch (Exception)
        {
            // Use the raw body when the server did not return structured JSON.
        }

        return new BubleApiException(statusCode, code, message, details, responseBody);
    }

    private Uri Resolve(string path, IDictionary<string, string> query)
    {
        var normalizedPath = path.StartsWith("/", StringComparison.Ordinal) ? path : "/" + path;
        var builder = new StringBuilder(BaseUrl).Append(normalizedPath);
        var hasQuery = normalizedPath.IndexOf("?", StringComparison.Ordinal) >= 0;

        foreach (var entry in query)
        {
            builder.Append(hasQuery ? '&' : '?');
            hasQuery = true;
            builder.Append(Uri.EscapeDataString(entry.Key));
            builder.Append('=');
            builder.Append(Uri.EscapeDataString(entry.Value));
        }

        return new Uri(builder.ToString(), UriKind.Absolute);
    }

    private static string TrimTrailingSlash(string? value)
    {
        var output = string.IsNullOrWhiteSpace(value) ? BubleClient.DefaultBaseUrl : value!;
        while (output.EndsWith("/", StringComparison.Ordinal))
        {
            output = output.Substring(0, output.Length - 1);
        }

        return output;
    }

    public void Dispose()
    {
        _httpClient.Dispose();
    }
}
